### PR TITLE
cors: clarify behavior when properties are defined on RouteOptions and VirtualHostOptions

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

Clarify the behavior of CORS policy when defined at multiple levels

# Context

The attached issue only is relevant due to the assumption that policies are merged. This assumption was incorrect, but was based off of documentation in our APIs.

This PR updates that documentation, and adds a test to demonstrate it.

## Testing steps
- Added an integration tests for envoy to demonstrate the behavior when RouteOptions and VirtualHostOptions are defined

## Notes for reviewers
-

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: test/e2e/cors_test.go [hunk 1]
```diff
@@ -51,144 +51,226 @@ var _ = Describe("CORS", func() {
...
+				}, "10s", ".1s").ShouldNot(HaveOccurred(), "Envoy config contains CORS filer")
+
+				allowedOriginRequestBuilder := testContext.GetHttpRequestBuilder().
+					WithOptionsMethod().
+					WithPath("cors").
+					WithHeader("Origin", allowedOrigins[0]).
+					WithHeader("Access-Control-Request-Method", http.MethodGet).
+					WithHeader("Access-Control-Request-Headers", "X-Requested-With")
+				Eventually(func(g Gomega) {
+					g.Expect(testutils.DefaultHttpClient.Do(allowedOriginRequestBuilder.Build())).Should(matchers.HaveOkResponseWithHeaders(map[string]interface{}{
```
Comment (jbohanon) [inline test/e2e/cors_test.go:99]: does this pop the bodyclose linter? might have to make the request, close the body, and assert the response, or add a `nolint`

Comment (jbohanon) [reply test/e2e/cors_test.go:99]: looks like it does not :+1:

**Review by nfuden — DISMISSED** [review]

File: changelog/v1.18.0-beta8/cors.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+changelog:
+  - type: FIX
+    issueLink: https://github.com/solo-io/gloo/issues/8115
+    resolvesIssue: false
+    description: >-
+      Clarify the inheritance behavior of CORS APIs defined on the VirtualHost and Route level. Previously,
+      there was a claim that when properties were defined at both levels, they were merged. However, the accurate
+      behavior is that RouteOptions entirely override VirtualHostOptions when both are defined.
```
Comment (jbohanon) [inline changelog/v1.18.0-beta8/cors.yaml:8]: Is this changelog comment still accurate?

Comment (nfuden) [inline changelog/v1.18.0-beta8/cors.yaml:8]:
```suggestion
    description: >-
      Clarify the inheritance behavior of CORS APIs defined on the VirtualHost and Route level. Previously,
      there was a claim that when properties were defined at both levels, they were merged. However, the accurate
      behavior is that RouteOptions fields entirely override VirtualHostOptions when RouteOptions are defined.
```

Comment (nfuden) [reply changelog/v1.18.0-beta8/cors.yaml]: Good point updated

**Review by nfuden — APPROVED** [review]

File: projects/gloo/api/v1/options.proto [hunk 3]
```diff
@@ -269,7 +269,8 @@ message VirtualHostOptions {
     // Append/Remove headers on Requests or Responses on all routes contained in this Virtual Host
     headers.options.gloo.solo.io.HeaderManipulation header_manipulation = 2;
     // Defines a CORS policy for the virtual host
-    // If a CORS policy is also defined on the route matched by the request, the policies are merged.
+    // If a CORS policy is also defined on the route matched by the request, the route policy
+    // overrides the virtual host policy for any configured field.
```
Comment (jenshu) [inline projects/gloo/api/v1/options.proto:273]: > the route policy overrides the virtual host policy for any configured field.

to clarify, does this mean the override will be done field-by-field, or that the entire route policy will override the entire vhost policy?

Comment (nfuden) [reply projects/gloo/api/v1/options.proto:273]: To be specific this is field by field

File: projects/gloo/api/v1/options.proto [hunk 4]
```diff
@@ -436,7 +437,7 @@ message RouteOptions {
     // If true and there is a host rewrite, appends the x-forwarded-host header to requests.
     google.protobuf.BoolValue append_x_forwarded_host = 146;
     // Defines a CORS policy for the route
-    // If a CORS policy is also defined on the route's virtual host, the policies are merged.
+    // If a CORS policy is also defined on the route's virtual host, it is ignored in favor of this policy
```
[the remark this answers is not in the document]
Comment (nfuden) [reply projects/gloo/api/v1/options.proto:440]: This language is incorrect.Will update it

File: changelog/v1.18.0-beta8/cors.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+changelog:
+  - type: FIX
+    issueLink: https://github.com/solo-io/gloo/issues/8115
+    resolvesIssue: false
+    description: >-
+      Clarify the inheritance behavior of CORS APIs defined on the VirtualHost and Route level. Previously,
+      there was a claim that when properties were defined at both levels, they were merged. However, the accurate
+      behavior is that RouteOptions fields entirely override VirtualHostOptions when RouteOptions are defined.
```
Comment (jenshu) [inline changelog/v1.18.0-beta8/cors.yaml:8]: i think this language should also be updated to clarify that it's field-by-field

> RouteOptions fields entirely override VirtualHostOptions when RouteOptions are defined

**Review by nfuden — APPROVED** [review]

**Review by jenshu — APPROVED** [review]

**Pull request merged**